benchmarks/profiler_string_storage_intern.rb

```diff
@@ -0,0 +1,40 @@
+# Used to quickly run benchmark under RSpec as part of the usual test suite, to validate it didn't bitrot
+VALIDATE_BENCHMARK_MODE = ENV['VALIDATE_BENCHMARK'] == 'true'
+
+return unless __FILE__ == $PROGRAM_NAME || VALIDATE_BENCHMARK_MODE
+
+require_relative 'benchmarks_helper'
+
+# This benchmark measures the performance of string storage intern/intern_all APIs
+
+class ProfilerStringStorageIntern
+  def initialize
+    @recorder = Datadog::Profiling::StackRecorder.for_testing(heap_samples_enabled: true)
+  end
+
+  def run_benchmark
+    Benchmark.ips do |x|
+      benchmark_time = VALIDATE_BENCHMARK_MODE ? { time: 0.01, warmup: 0 } : { time: 10, warmup: 2 }
+      x.config(
+        **benchmark_time,
+      )
+
+      x.report('intern 1000 strings') do
+        Datadog::Profiling::StackRecorder::Testing._native_benchmark_intern(@recorder, "hello, world!", 1000, false)
+      end
+
+      x.report('intern_all 1000 strings') do
```

Comment:
Not a blocker but there may be value in benchmarking the "intern new string" path? Both of these benchmarks are really only exercising the "existing string" path?

And maybe having a realistic balance based on some statistics of profiling on typical Ruby apps (e.g. 90% new strings, 10% re-use).

Reply:
TBH I added this benchmark to motivate the work on https://github.com/DataDog/libdatadog/pull/844 .

I'm not quite sure testing other mixes of strings would make a lot of difference in comparing `intern` vs `intern_all`, since in the end both call the string table in the same way, it's just that the latter avoids crossing the FFI and acquiring the write lock many times.

Or were you thinking more along the lines of turning this benchmark into a measurement of the string table performance in general? So that, e.g. if there was a regression we'd catch it?

Reply:
Yeah without that context I had assumed the intent was to more easily catch regressions or understand evolution of performance of this part.

Reply:
I like your suggestion: In c29be53269ebe17db3f226b8c662200ad3f2f49a I've modified the benchmark to track performance of managed string storage more generally, rather than just comparing intern vs intern_all.